src/GraphQl/Type/FieldsBuilder.php

```diff
@@ -0,0 +1,371 @@
+                if (!isset($value['#name'])) {
+                    $name = (false === $pos = strrpos($original, '[')) ? $original : substr($original, 0, (int) $pos);
+                    $value['#name'] = ($resourceMetadata ? $resourceMetadata->getShortName() : '').'Filter_'.strtr($name, ['[' => '_', ']' => '', '.' => '__']);
+                }
+            }
+
+            $args[$key] = $value;
+        }
+
+        return $args;
+    }
+
+    private function convertFilterArgsToTypes(array $args): array
+    {
+        foreach ($args as $key => $value) {
+            if (strpos($key, '.')) {
+                // Declare relations/nested fields in a GraphQL compatible syntax.
+                $args[str_replace('.', '_', $key)] = $value;
+                unset($args[$key]);
+            }
+        }
+
+        foreach ($args as $key => $value) {
+            if (!\is_array($value) || !isset($value['#name'])) {
+                continue;
+            }
+
+            $name = $value['#name'];
+
+            if ($this->typesContainer->has($name)) {
+                $args[$key] = $this->typesContainer->get($name);
+                continue;
+            }
+
+            unset($value['#name']);
+
+            $filterArgType = new InputObjectType([
+                'name' => $name,
+                'fields' => $this->convertFilterArgsToTypes($value),
+            ]);
+
+            $this->typesContainer->set($name, $filterArgType);
+
+            $args[$key] = $filterArgType;
+        }
+
+        return $args;
+    }
+
+    /**
+     * Converts a built-in type to its GraphQL equivalent.
+     *
+     * @throws InvalidTypeException
+     */
+    private function convertType(Type $type, bool $input, ?string $queryName, ?string $mutationName, string $resourceClass, ?string $property, int $depth)
+    {
+        $graphqlType = $this->typeConverter->convertType($type, $input, $queryName, $mutationName, $resourceClass, $property, $depth);
+
+        if (null === $graphqlType) {
+            throw new InvalidTypeException(sprintf('The type "%s" is not supported.', $type->getBuiltinType()));
+        }
+
+        if (\is_string($graphqlType)) {
+            if (!$this->typesContainer->has($graphqlType)) {
+                throw new InvalidTypeException(sprintf('The GraphQL type %s is not valid. Valid types are: %s. Have you registered this type by implementing %s?', $graphqlType, implode(', ', array_keys($this->typesContainer->all())), TypeInterface::class));
+            }
+
+            $graphqlType = $this->typesContainer->get($graphqlType);
+        }
+
+        if ($this->typeBuilder->isCollection($type)) {
+            return $this->paginationEnabled && !$input ? $this->typeBuilder->getResourcePaginatedCollectionType($graphqlType) : GraphQLType::listOf($graphqlType);
+        }
+
+        return $type->isNullable() || (null !== $mutationName && 'update' === $mutationName) ? $graphqlType : GraphQLType::nonNull($graphqlType);
```

Comment:
Why not just `return $graphqlType;`?

I did a fork because of this strange condition (to remove it).